### PR TITLE
backport plugin bundling to 2.2

### PR DESCRIPTION
**Pull request by rmuir**
This is the backport of #15233 but for maven.

It has the same filesystem layout and semantics.

The way it works in maven is, most plugin logic moves to modules/pom.xml, and plugin is a child of that, only it overrides integ logic to call install-plugin instead of install-module.

distributions for each packaging format (except integ-zip) basically just unpack modules into their modules/ directory.

The oddities/hacks/etc needed (IMO these are minor and such things are expected):
- Modules pom has a unnecessary project-level dependency on the integ-zip, that forces proper build order in the "nuclear reactor".
- Distribution pom requires distributions to "opt-in" to modules by enabling them for their specific distro. This way it itself does not depend on them and the integ-zip works, no loops, no downloading.
- Licensing works different than gradle, we inspect packaging here for the check. our module.jars are not third party dependencies: we have to generate an ignore pattern for them. 
- Licensing for modules gets "duplicated": its specified in the module itself and in the distribution. We could disable it for modules, but since I think publishing artifacts for these is unavoidable with maven, and since the license-checker is packaging-based, its correct.

### CODE REVIEW

**Comment by nik9000** [comment]
I've added this to the top of my review.

**Comment by nik9000** [comment]
> Modules pom has a unnecessary project-level dependency on the integ-zip, that forces proper build order in the "nuclear reactor".
> Distribution pom requires distributions to "opt-in" to modules by enabling them for their specific distro. This way it itself does not depend on them and the integ-zip works, no loops, no downloading.
> Licensing works different than gradle, we inspect packaging here for the check. our module.jars are not third party dependencies: we have to generate an ignore pattern for them.
> Licensing for modules gets "duplicated": its specified in the module itself and in the distribution. We could disable it for modules, but since I think publishing artifacts for these is unavoidable with maven, and since the license-checker is packaging-based, its correct.

These are all hacks that I expected to get this working in maven.

File: core/src/main/java/org/elasticsearch/plugins/PluginManager.java [hunk 1]
```diff
@@ -219,6 +221,12 @@ private void extract(PluginHandle pluginHandle, Terminal terminal, Path pluginFi
         PluginInfo info = PluginInfo.readFromProperties(root);
         terminal.println(VERBOSE, "%s", info);
 
+        // don't let luser install plugin as a module...
```
Comment (nik9000) [inline core/src/main/java/org/elasticsearch/plugins/PluginManager.java]: opinionated type still persists.

**Comment by nik9000** [comment]
LGTM. Thanks for the tour of the hacks.

**Comment by rmuir** [comment]
Thanks for looking this over @nik9000 !

**Pull request merged**